Use Case: Dynamic config w/Vault Dynamic secrets

paladin-devops opened the pull request:
The Use Case in this PR goes over how one can configure Vault and Waypoint to use dynamic secrets in their dynamic application configuration, with the help of config sourcers and the Waypoint Entrypoint.

Note: the capability of using dynamic secrets from Vault depends on a fix to the Vault plugin that is yet to be released in v0.10.4.

paladin-devops committed:
docs: Use case for Waypoint dynamic config w/Vault dynamic secrets.

briancain commented on website/content/docs/intro/use-cases/dynamic-config-vault-dynamic-secrets.mdx:
```diff
@@ -0,0 +1,269 @@
+
+# Dynamic Secrets for Waypoint with Vault
+
+Hashicorp Waypoint is an application deployment platform that simplifies how users
```
```suggestion
HashiCorp Waypoint is an application deployment platform that simplifies how users
```

briancain commented on website/content/docs/intro/use-cases/dynamic-config-vault-dynamic-secrets.mdx:
```diff
@@ -0,0 +1,269 @@
+or files can be applied to your running deployment. Alternatively, configuration
+can be sourced dynamically from external sources, using a config sourcer plugin.
+
+The [Hashicorp Vault plugin for Waypoint](https://developer.hashicorp.com/waypoint/plugins/vault)
```
```suggestion
The [HashiCorp Vault plugin for Waypoint](https://developer.hashicorp.com/waypoint/plugins/vault)
```

briancain commented on website/content/docs/intro/use-cases/dynamic-config-vault-dynamic-secrets.mdx:
```diff
@@ -0,0 +1,269 @@
+that is built-in to the Waypoint binary. This plugin can be used to source static
+values from the Vault KV secrets engine, but also for dynamic secrets! A [dynamic secret](https://developer.hashicorp.com/vault/tutorials/getting-started/getting-started-dynamic-secrets)
+is one that is managed by Vault and has a lease associated with it, where
+the secret will expire after a set time to live (TTL). This is a huge benefit to
```
```suggestion
the secret will expire after a set time to live (TTL). This is a huge benefit for
```

briancain commented on website/content/docs/intro/use-cases/dynamic-config-vault-dynamic-secrets.mdx:
```diff
@@ -0,0 +1,269 @@
+expire after a pre-configured amount of time, among other things including that
+developers and operators don’t need to know secrets for their deployed applications.
+
+Prior to using Waypoint to deploy my Nomad jobs, I’ve relied on [Nomad’s integration
```
A quick note on this paragraph, since this isn't a blog post with an Author attached to it, the voice here should be less "I" "My" because the reader won't really know who that is.

For this example specifically, you could rephrase it to say like

> Prior to Waypoint, Nomad users would have to rely on Nomads integration with Vault....

But this applies to the rest of this post too!

paladin-devops replied:
Makes sense. Since this section is new to the site, _"I"_ wasn't sure which way to go with it, but I'll undo the 1st person.

briancain commented on website/content/docs/intro/use-cases/dynamic-config-vault-dynamic-secrets.mdx:
```diff
@@ -0,0 +1,269 @@
+when I’m not deploying to my Nomad cluster. Let’s look into how I configured this
+in Waypoint.
+
+My application, located at [this path](https://github.com/hashicorp/waypoint-examples/tree/dynamic-config-example/learn/dynamic-application-configuration)
```
For safety sake, let's point this at a specific commit rather than `main` in case the link goes stale

paladin-devops replied:
I agree - left this on my separate branch since this change isn't merged to `main` yet, but I'll open a PR on the examples repo for that pretty soon!

briancain commented on website/content/docs/intro/use-cases/dynamic-config-vault-dynamic-secrets.mdx:
````diff
@@ -0,0 +1,269 @@
+ready by adding it to our waypoint.hcl file, and committing it to the git repository
+for the project.
+
+```hcl
````
You might want to copy this out, and run a `waypoint fmt` on it so everything is aligned!

paladin-devops replied:
I did this many a time in the Google Doc, and lost some formatting when converting to PR 😅 will do!

briancain commented on website/content/docs/intro/use-cases/dynamic-config-vault-dynamic-secrets.mdx:
````diff
@@ -0,0 +1,269 @@
+}
+```
+
+In the configuration above, we’ve informed Waypoint that we want the Waypoint Entrypoint,
````
Maybe turn `Waypoint Entrypoint` into a URL to link to what that is in the docs?

jgwhite replied:
I wonder if we can avoiding introducing the concept of the CEB altogether here. It’s pretty advanced stuff and may leave the reader feeling overloaded with concepts. Loose suggestion:

```markdown
In the configuration above, we’ve informed Waypoint that we want to get a
secret from Vault at the path `database/creds/my-role` and make it available in
our app’s container. To learn more about how Waypoint does this, see the
[Waypoint Entrypoint docs](https://developer.hashicorp.com/waypoint/docs/entrypoint).
```

briancain commented on website/content/docs/intro/use-cases/dynamic-config-vault-dynamic-secrets.mdx:
```diff
@@ -0,0 +1,269 @@
+can now dynamically source this Vault configuration wherever the app is deployed.
+If I deploy to Docker on my laptop for some local development testing, or my EKS
+cluster in AWS, as long as the application container can access the Vault cluster,
+the secrets necessary to run the application will be provided by Waypoint.
```
Maybe wrap this up with further reading, learn guide links, links to other docs for more info, etc?

briancain reviewed (changes requested):
Very dope! Well done. I have a few edits for suggestions, but overall the content looks great. Excited to get this on the website!

jgwhite commented on website/content/docs/intro/use-cases/dynamic-config-vault-dynamic-secrets.mdx:
```diff
@@ -0,0 +1,269 @@
+
+# Dynamic Secrets for Waypoint with Vault
+
+Hashicorp Waypoint is an application deployment platform that simplifies how users
+test and release apps onto platforms and utilize infrastructure managed by Terraform.
```
Do we need this preamble, or can we assume readers will already have a basic grasp of what Waypoint does?

Comparing with [Vault’s use-case tutorials](https://developer.hashicorp.com/vault/tutorials/db-credentials/database-secrets), it seems like they dive straight into talking about the feature in question, rather than setting the stage with a general pitch for the product.

paladin-devops replied:
I think this can be removed! I included it at the beginning when I was drafting this as a blog post for a quick intro, but since this is embedded in our website I think we can assume the reader went through the introduction first. 😄

jgwhite commented on website/content/docs/intro/use-cases/dynamic-config-vault-dynamic-secrets.mdx:
````diff
@@ -0,0 +1,269 @@
+connect to Vault. A configuration source for the Vault plugin will be added like so:
+
+```shell
+$ waypoint config source-set -type=vault -config=addr=https://vault.service.consul:8200 -config=token=<VAULT_TOKEN>
````
For the purposes of this use-case/tutorial should we assume the reader is running Vault with `vault server -dev`?

```suggestion
$ waypoint config source-set -type=vault -config=addr=http://127.0.0.1:8200 -config=token=<VAULT_TOKEN>
```

paladin-devops replied:
I think that's fair, I originally had this as localhost, as you've depicted it here, but thought to try using what could perhaps be a "production" Vault URL, but I don't feel strongly about it. I'll update it!

paladin-devops committed:
Apply suggestions from code review

Co-authored-by: Brian Cain <bcain@hashicorp.com>
Co-authored-by: Jamie White <jamie@jgwhite.co.uk>

paladin-devops committed:
docs: Update Vault dynamic secrets use case.

Change point-of-view from which the use case is written, add links for further reading, and other minor updates.

paladin-devops committed:
docs: Update Vault dynamic secrets use case link to examples repo.

Update link to route to a specific commit instead of a branch.

paladin-devops committed:
docfix: Vault use case format.

paladin-devops committed:
docs: List items for further reading in Vault use case.

paladin-devops committed:
docs: Minor updates to Vault config sourcer use case.
